src/Language/Fixpoint/Solver/Interpreter.hs

```diff
@@ -0,0 +1,887 @@
+  where
+    res'     = M.fromList $ zip is ps''
+    ps''     = zipWith (\i -> elaborate (atLoc dummySpan ("PLE1 " ++ show i)) env) is ps' 
+    ps'      = defuncAny cfg env ps
+    (is, ps) = unzip (M.toList res)
+
+---------------------------------------------------------------------------------------------- 
+-- | @InstEnv@ has the global information needed to do PLE
+---------------------------------------------------------------------------------------------- 
+
+data InstEnv a = InstEnv 
+  { ieCfg   :: !Config
+  , ieBEnv  :: !BindEnv
+  , ieAenv  :: !AxiomEnv 
+  , ieCstrs :: ![(SubcId, SimpC a)]
+  , ieKnowl :: !Knowledge
+  , ieEvEnv :: !EvalEnv
+  } 
+
+---------------------------------------------------------------------------------------------- 
+-- | @ICtx@ is the local information -- at each trie node -- obtained by incremental PLE
+---------------------------------------------------------------------------------------------- 
+
+data ICtx    = ICtx 
+  { icAssms    :: S.HashSet Pred            -- ^ Equalities converted to SMT format
+  , icCands    :: S.HashSet Expr            -- ^ "Candidates" for unfolding
+  , icEquals   :: EvAccum                   -- ^ Accumulated equalities
+  , icSolved   :: S.HashSet Expr            -- ^ Terms that we have already expanded
+  , icSimpl    :: !ConstMap                 -- ^ Map of expressions to constants
+  , icSubcId   :: Maybe SubcId              -- ^ Current subconstraint ID
+  } 
+
+---------------------------------------------------------------------------------------------- 
+-- | @InstRes@ is the final result of PLE; a map from @BindId@ to the equations "known" at that BindId
+---------------------------------------------------------------------------------------------- 
+
+type InstRes = M.HashMap BindId Expr
+
+---------------------------------------------------------------------------------------------- 
+-- | @Unfold is the result of running PLE at a single equality; 
+--     (e, [(e1, e1')...]) is the source @e@ and the (possible empty) 
+--   list of PLE-generated equalities (e1, e1') ... 
+---------------------------------------------------------------------------------------------- 
+
+type CTrie   = T.Trie   SubcId
+type CBranch = T.Branch SubcId
+type Diff    = [BindId]    -- ^ in "reverse" order
+
+initCtx :: InstEnv a -> [(Expr,Expr)] -> ICtx
+initCtx _   es   = ICtx 
+  { icAssms  = mempty 
+  , icCands  = mempty 
+  , icEquals = S.fromList es
+  , icSolved = mempty
+  , icSimpl  = mempty 
+  , icSubcId = Nothing
+  }
+
+equalitiesPred :: S.HashSet (Expr, Expr) -> [Expr]
+equalitiesPred eqs = [ EEq e1 e2 | (e1, e2) <- S.toList eqs, e1 /= e2 ] 
+
+updCtxRes :: InstRes -> Maybe BindId -> ICtx -> (ICtx, InstRes) 
+updCtxRes res iMb ctx = (ctx, res')
+  where 
+    res' = updRes res iMb (pAnd $ equalitiesPred $ icEquals ctx)
+
+
+updRes :: InstRes -> Maybe BindId -> Expr -> InstRes
+updRes res (Just i) e = M.insert i e res 
```

Comment:
This algorithm seems to depend on the tree-like invariant too. Here it could check it as done in https://github.com/ucsd-progsys/liquid-fixpoint/pull/498/commits/a0f2dc95ed9b3c7606ceea5a17db97ce40c08048

Reply:
Perhaps it is worth factoring the common functions so the code isn't duplicated for the sake of this changes.

Reply:
> This algorithm seems to depend on the tree-like invariant too. Here it could check it as done in a0f2dc9

This has been addressed.